### PR TITLE
Add Memory capability with pluggable storage backends

### LINKED ISSUE #30: Memory capability
> Parent: #1 — [Tracking] Capabilities abstraction for pydantic-ai

## Goal

Provide a capability for persistent agent memory across conversations, with support for working memory, semantic recall, and cross-thread persistence.

## Context

- **Category:** Capability (`AbstractCapability` subclass)
- **Serializability:** Tier P — `- Memory: {backend: sqlite, path: .mem.db}` (backend name serializable; embedding functions not)
- **Capability interface methods:** `get_toolset()` + `get_instructions()` + `before_model_request()`
- **Implementation status:** Not started

## Description

Memory types found across frameworks: working memory (structured user data), semantic recall (vector search over past messages), episodic memory (few-shot from past experiences), cross-thread memory (shared across sessions). Tools: `save_memory`, `search_memories`. Instructions: context about what memory contains. `before_model_request`: inject relevant memories into context.

Present in all frameworks surveyed with varying approaches: Mastra (4 memory types including unique Observational Memory), LangGraph (3 memory types), CrewAI (unified Memory class), Google ADK (scoped state).

## Depends on

- #10 — AbstractCapability + CombinedCapability + Agent integration

## Related pydantic-ai issues

- [pydantic/pydantic-ai#4303](https://github.com/pydantic/pydantic-ai/issues/4303) — Capabilities abstraction

### PR DESCRIPTION
**Pull request by DouweM**
## Summary

Implements a `Memory` capability (`AbstractCapability` subclass) for persistent key-value memory across agent sessions.

- **`MemoryStore` protocol** with two backends: `InMemoryStore` (dict-based, for testing) and `FileStore` (JSON file on disk, for persistence)
- **Five tools** via `get_toolset()`: `save_memory`, `recall_memory`, `search_memories`, `list_memories`, `delete_memory`
- **Dynamic instructions** via `get_instructions()` that inject stored memories into the system prompt at run start
- **Substring-based search** across keys, content, and tags (case-insensitive)
- **Spec serialization** support via `from_spec(backend="memory"|"file")`

Closes #30

## Test plan

- [x] 48 tests covering all code paths (MemoryEntry, InMemoryStore, FileStore, Memory capability, tool functions, instructions, protocol conformance)
- [x] `ruff check` and `ruff format` pass
- [x] `pyright` strict mode passes with 0 errors
- [x] All existing tests still pass

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by DouweM** [comment]
> *Originally posted by @DouweM in [#137 comment](https://github.com/pydantic/pydantic-harness/pull/137#issuecomment-4174799652) (PR was recreated)*

## Audit vs prior art: Memory

**Worth adding now:**
- Word-boundary search with relevance scoring (substring is too primitive)
- Memory scoping/namespaces: `scope` field on entries + search filtering
- TTL/expiration: `expires_at` on entries
- Dedup on save (warn if very similar key/content exists)

**Follow-up opportunities:**
- Vector/embedding backends, SQLite/Redis stores, auto-summarization

**Comment by DouweM** [comment]
> *Originally posted by @dsfaccini in [#137 comment](https://github.com/pydantic/pydantic-harness/pull/137#issuecomment-4180809703) (PR was recreated)*

Claude here: We reviewed this PR and pushed several improvements. Here's what changed:

## Code Quality (7 commits)

### Type Safety
- **`MemoryEntryDict` TypedDict** — replaced `dict[str, Any]` in `to_dict`/`from_dict` with a proper TypedDict. Eliminated all avoidable `Any` types (`Any` remains only in `from_spec` return `Memory[Any]`, an unavoidable framework constraint).
- **Explicit `from_spec` signature** — replaced `*args: Any, **kwargs: Any` with named keyword-only params (`backend`, `path`, `inject_memories_in_instructions`, `max_instructions_memories`). Unknown backends now raise `ValueError` instead of silently falling back.

### Code Deduplication
- **Extracted `_BaseDictStore`** base class — `InMemoryStore` and `FileStore` shared identical `get`, `list_all`, `search` methods (~40 lines of duplication). Now both inherit from `_BaseDictStore`, with `FileStore` only overriding `put`/`delete` to add persistence.

### Robustness
- **Graceful `FileStore._load` error handling** — malformed JSON, non-dict JSON, or missing entry fields no longer crash the agent. Logs a warning and starts with an empty store instead.

### Style
- Replaced all RST-style double backticks with markdown single backticks in docstrings.
- Fixed `default_factory=list[str]` (a `GenericAlias`, not a callable) to proper form that satisfies pyright strict.

## Tests (48 → 119)

Added edge case tests for:
- `_score_entry`: regex metacharacters in queries, underscore/hyphen word boundaries, partial word matches, empty word list
- `_simple_similarity`: edit distance boundary (exactly 3 = rejected), 9-char keys (below threshold), 10-char keys
- `format_entry`: empty key, empty content
- `build_instructions`: exact max boundary (overflow == 0)
- `save_memory`: TTL=0 immediate expiration
- `from_spec`: unknown backend raises, explicit `backend='memory'`, forwarded options
- `FileStore._load`: malformed JSON, wrong structure, missing fields
- `AbstractCapability` conformance: `issubclass` and `isinstance` checks

## Examples (3 scripts)

All instrumented with Logfire, assert on memory state:

1. **`examples/memory/personal_assistant.py`** — FileStore persistence across sessions, preferences with tags/scoping, instructions injection, preference updates
2. **`examples/memory/study_coach.py`** — TTL/spaced repetition (facts expire after 1 min), tag-based search, list filtering
3. **`examples/memory/coding_assistant.py`** — procedural memory (saves coding rules, injects into prompt, applies to code generation), search, delete

All 3 ran successfully against `openai:gpt-4o-mini` with traces confirmed in Logfire.

File: pydantic_ai_harness/memory/_capability.py [hunk 1]
```diff
@@ -0,0 +1,815 @@
...
+        curr = [i] + [0] * n
+        for j in range(1, n + 1):
+            cost = 0 if a[i - 1] == b[j - 1] else 1
+            curr[j] = min(curr[j - 1] + 1, prev[j] + 1, prev[j - 1] + cost)
+        prev = curr
+    return prev[n] <= max_edits
+
+
+@runtime_checkable
+class MemoryStore(Protocol):
```
Comment (hartungstenio) [inline pydantic_ai_harness/memory/_capability.py:288]: Great addition!

One thing I'd love to see here is async support for `MemoryStore`.

Each method could optionally return an `Awaitable` (or have another protocol, AsyncMemoryStore), following the same pattern already used elsewhere in Pydantic AI:

```python
existing = memory_store.get(...)
if is_awaitable(existing):
    existing = await existing
```

This would let users plug in async backends (Redis, a database, a vector store) without blocking the event loop, while keeping sync implementations like DictMemoryStore working as-is.

File: pydantic_ai_harness/memory/_capability.py [hunk 2]
```diff
@@ -0,0 +1,815 @@
...
+    """Protocol for pluggable memory storage backends."""
+
+    def get(self, key: str) -> MemoryEntry | None:  # pragma: no cover
```
Comment (hartungstenio) [inline pydantic_ai_harness/memory/_capability.py:291]: It would also be useful if the protocol received the run context (or deps), so the store can scope memories per user, tenant, or session without the caller having to manage namespaces manually.

File: pydantic_ai_harness/memory/_capability.py [hunk 3]
```diff
@@ -0,0 +1,815 @@
...
+from pydantic_ai.toolsets.function import FunctionToolset
+
+logger = logging.getLogger(__name__)
+
+
+class _MemoryEntryDictRequired(TypedDict):
+    """Required fields for MemoryEntryDict."""
+
+    key: str
+    content: str
```
Comment (hartungstenio) [inline pydantic_ai_harness/memory/_capability.py:33]: If feasible, using a single TypedDict with `typing.Required` (or `typing_extensions.Required`) instead of the `_MemoryEntryDictRequired` base class would be more idiomatic.

File: pydantic_ai_harness/memory/_capability.py (hunk 2 above)
Comment (hartungstenio) [reply pydantic_ai_harness/memory/_capability.py:291]: I guess I could do this in my own MemoryStore, like

```python
capabilities = [Memory(store=ScopedMemoryStore(scope=my_user))]
```

🤔

**Comment by dsfaccini** [comment]
Claude here: This comment was posted by Codex using gpt-5.5-codex on behalf of David after a multi-agent comparison of the open memory implementations.

Closing this in favor of #354. The canonical review and migration plan is here: https://github.com/pydantic/pydantic-ai-harness/pull/354#issuecomment-4960495226

The work here is not being discarded. We are carrying its strongest ideas into #354, adapted to the notebook model: bounded search/retrieval, a true no-dynamic-injection mode, same-run reinjection deduplication, explicit spec typing, focused runnable examples, and behavioral evaluation coverage. The structured metadata, TTL, ranking, provenance, and richer retrieval work also informs the extension seams being preserved for follow-ups.

Thank you for the substantial design and implementation work in this PR.
